meson.build

```diff
@@ -0,0 +1,237 @@
+  message('Disabling FORTIFY_SOURCE as optimization is set to 0')
+endif
+
+cmake = import('cmake')
+pkg = import('pkgconfig')
+
+
+# init configuration
+#-------------------------------------------------
+config_h = configuration_data()
+
+config_h.set(
+    'HAVE_CXX17', 1,
+    description : 'Define if the compiler supports basic C++17 syntax')
+
+package_name = meson.project_name()
+config_h.set_quoted(
+    'PACKAGE_NAME', package_name,
+    description : 'Define to the full name of this package.'
+    )
+
+package_tarname = meson.project_name()
+config_h.set_quoted(
+    'PACKAGE_TARNAME', package_tarname,
+    description : 'Define to the one symbol short name of this package.')
+
+package_version = meson.project_version()
+config_h.set_quoted(
+    'PACKAGE_VERSION', package_version,
+    description : 'Define to the version of this package.')
+
+package_string = '@0@ @1@'.format(package_name, package_version)
+config_h.set_quoted(
+    'PACKAGE_STRING', package_string,
+    description : 'Define to the full name and version of this package.')
+
+package_url = 'https://nixos.org/nix/'
+config_h.set_quoted(
+    'PACKAGE_URL', package_url,
+    description : 'Define to the home page for this package.')
+
+package_bug_url = 'https://github.com/nixos/nix/issues'
+config_h.set_quoted(
+    'PACKAGE_BUGREPORT', package_bug_url,
+    description : 'Define to the address where bug reports for this package should be sent.')
+
+
+# env
+#============================================================================
+
+
+# set install directories
+#-------------------------------------------------
+prefix = get_option('prefix')
+libdir = join_paths(prefix, get_option('libdir'))
+bindir = join_paths(prefix, get_option('bindir'))
+datadir = join_paths(prefix, get_option('datadir'))
+sysconfdir = join_paths(prefix, get_option('sysconfdir'))
+libexecdir = join_paths(prefix, get_option('libexecdir'))
+mandir = join_paths(prefix, get_option('mandir'))
+docdir = join_paths(prefix, get_option('docdir'))
+includedir = join_paths(prefix, get_option('includedir'))
+proj_inc = []
+
+# set nix directories
+#-------------------------------------------------
+
+# State should be stored in /nix/var, unless the user overrides it explicitly.
+if get_option('normal_var')
+    localstatedir = '/nix/var'
+else
+    localstatedir =  join_paths(prefix, get_option('localstatedir'))
+endif
+
+nixstoredir = get_option('nixstoredir')
+
+profiledir =  join_paths(sysconfdir, 'profile.d')
+
+
+# Construct a Nix system name (like "i686-linux").
+#-------------------------------------------------
+machine_name = host_machine.cpu()
+sys_name = host_machine.system().to_lower()
+
+cpu_archs = ['x86_64', 'armv6', 'armv7', '']
+
+foreach cpu : cpu_archs 
+    if host_machine.cpu().contains(cpu)
+        if cpu.contains('armv')
+            machine_name = cpu + '1'
+        else
+            machine_name = cpu
+        endif
+        break
+    endif
+endforeach
```

Comment:
The only information I could find about `host_machine.cpu()` is that it's ["very system dependent"](https://mesonbuild.com/Cross-compilation.html#machine-entries), which obviously offers no compatibility guarantees with the current scheme which depends on [`autoconf` (and `config/config.guess`)](https://www.gnu.org/software/autoconf/manual/autoconf-2.67/autoconf.html#Hosts-and-Cross_002dCompilation).

I mention this because the system naming scheme is heavily depended upon, [was crystallized in the platform support tier RFC](https://github.com/NixOS/rfcs/blob/master/rfcs/0046-platform-support-tiers.md), is part of the (unstable) flakes schema.
It therefore has become an interface that should not be broken for no reason, regardless if that was intended or not.

I think that any port to another build system would have to use `config/config.guess` to ensure compatibility.

It might be valuable to have a normalized naming scheme that didn't depend on the build system, but it would have to come through a non-breaking transition period (using `experimental-features` perhaps) and finally a major Nix update.

I think the logic in the `foreach` might also be buggy (will produce either `x86_64`, `armv61`, `armv71` or ``), but that's a minor point.

Reply:
We have CPU and cpu_family. Cpu_family is guaranteed to not change once value is documented. cpu is meant to be used to enable microarch specific optimizations. https://mesonbuild.com/Reference-tables.html#cpu-families

If there's a need to detect more specific build cpu info we can likely add a feature for that. When cross compiling you control what's in that value all the time through the cross file

Reply:
I'm aware of `cpu_family()` but it also wouldn't work by itself because, namely, there's no distinction between sub-architectures for Arm. I suppose that a combination of `cpu_family()` plus filling in what's needed by inspecting `cpu()`
Here's a sample of the `arch-os` doubles people have recorded in `nixpkgs` so far: https://github.com/NixOS/nixpkgs/blob/master/lib/systems/doubles.nix

Reply:
I think that could all be constructed through the `build_machine` and `host_machine` object, if there was a way to set the `cpu()` for the `build_machine` object. I took a quick glance at that list, and everything could be built from a mixture of `cpu_family()`, `system()` and `endian()` except for some microarches of arm. @Ericson2314 and I have talked about the insufficiencies of the various `machines` in the past, especially for nix, so there's room for improvement here for sure.

Reply:
I went ahead and opened https://github.com/mesonbuild/meson/issues/9720, to discuss adding a mechanism for overriding the `build_machine.cpu()` (in a host == build configuration the `host_machine` object is the `build_machine` object). I'm opened to other ideas as well.